# N-Gram models
>  Learn about n-gram modeling and use it to perform sentiment analysis on movie reviews.

- toc: true 
- badges: true
- comments: true
- author: Lucas Nunes
- categories: [Datacamp]
- image: images/datacamp/___

> Note: This is a summary of the course's chapter 3 exercises "Feature Engineering for NLP in Python" at datacamp. <br>[Github repo](https://github.com/lnunesAI/Datacamp/) / [Course link](https://www.datacamp.com/tracks/machine-learning-scientist-with-python)

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import spacy

plt.rcParams['figure.figsize'] = (8, 8)

## Building a bag of words model

### Word vectors with a given vocabulary

<div class=""><p>You have been given a corpus of documents and you have computed the vocabulary of the corpus to be the following:
<strong><em>V</em></strong>: <em>a, an, and, but, can, come, evening, forever, go, i, men, may, on, the, women</em></p>
<p>Which of the following corresponds to the bag of words vector for the document "men may come and men may go but i go on forever"?</p></div>

<pre>
Possible Answers

<b>(0, 0, 1, 1, 0, 1, 0, 1, 2, 1, 2, 2, 1, 0, 0)</b>

(0, 1, 0, 1, 1, 1, 2, 0, 2, 1, 0, 0, 0, 2, 0)

(2, 1, 0, 0, 2, 1, 0, 0, 0, 1)

(0, 0, 1, 2, 1, 2, 1, 1, 1, 0, 0, 1, 1, 1, 1)

</pre>

**That is, indeed, the correct answer. Each value in the vector corresponds to the frequency of the corresponding word in the vocabulary.**

### BoW model for movie taglines

<div class=""><p>In this exercise, you have been provided with a <code>corpus</code> of more than 7000 movie tag lines. Your job is to generate the bag of words representation <code>bow_matrix</code> for these taglines. For this exercise, we will ignore the text preprocessing step and generate <code>bow_matrix</code> directly.</p>
<p>We will also investigate the shape of the resultant <code>bow_matrix</code>. The first five taglines in <code>corpus</code> have been printed to the console for you to examine.</p></div>

In [ ]:
df = pd.read_csv('https://raw.githubusercontent.com/lnunesAI/Datacamp/main/2-machine-learning-scientist-with-python/13-Feature%20Engineering%20for%20NLP%20in%20Python/datasets/movie_overviews.csv').dropna()
corpus = df['tagline']

Instructions
<ul>
<li>Import the <code>CountVectorizer</code> class from <code>sklearn</code>.</li>
<li>Instantiate a <code>CountVectorizer</code> object. Name it <code>vectorizer</code>.</li>
<li>Using <code>fit_transform()</code>, generate <code>bow_matrix</code> for <code>corpus</code>.</li>
</ul>

In [ ]:
# Import CountVectorizer
from sklearn.feature_extraction.text import CountVectorizer

# Create CountVectorizer object
vectorizer = CountVectorizer()

# Generate matrix of word vectors
bow_matrix = vectorizer.fit_transform(corpus)

# Print the shape of bow_matrix
print(bow_matrix.shape)

(7033, 6614)


**You now know how to generate a bag of words representation for a given corpus of documents. Notice that the word vectors created have more than 6600 dimensions. However, most of these dimensions have a value of zero since most words do not occur in a particular tagline.**

### Analyzing dimensionality and preprocessing

<div class=""><p>In this exercise, you have been provided with a <code>lem_corpus</code> which contains the pre-processed versions of the movie taglines from the previous exercise. In other words, the taglines have been lowercased and lemmatized, and stopwords have been removed.  </p>
<p>Your job is to generate the bag of words representation <code>bow_lem_matrix</code> for these lemmatized taglines and compare its shape with that of <code>bow_matrix</code> obtained in the previous exercise. The first five lemmatized taglines in <code>lem_corpus</code> have been printed to the console for you to examine.</p></div>

In [ ]:
lem_corpus = pd.read_csv('https://github.com/lnunesAI/Datacamp/raw/main/2-machine-learning-scientist-with-python/13-Feature%20Engineering%20for%20NLP%20in%20Python/datasets/lem_corpus.csv')['1']



---



Instructions
<ul>
<li>Import the <code>CountVectorizer</code> class from <code>sklearn</code>.</li>
<li>Instantiate a <code>CountVectorizer</code> object. Name it <code>vectorizer</code>.</li>
<li>Using <code>fit_transform()</code>, generate <code>bow_lem_matrix</code> for <code>lem_corpus</code>.</li>
</ul>

In [ ]:
# Create CountVectorizer object
vectorizer = CountVectorizer()

# Generate matrix of word vectors
bow_lem_matrix = vectorizer.fit_transform(lem_corpus)

# Print the shape of bow_lem_matrix
print(bow_lem_matrix.shape)

(6959, 5223)


**Notice how the number of features have reduced significantly from around 6600 to around 5223 for pre-processed movie taglines. The reduced number of dimensions on account of text preprocessing usually leads to better performance when conducting machine learning and it is a good idea to consider it. However, as mentioned in a previous lesson, the final decision always depends on the nature of the application.**

### Mapping feature indices with feature names

<div class=""><p>In the lesson video, we had seen that <code>CountVectorizer</code> doesn't necessarily index the vocabulary in alphabetical order. In this exercise, we will learn to map each feature index to its corresponding feature name from the vocabulary.</p>
<p>We will use the same three sentences on lions from the video. The sentences are available in a list named <code>corpus</code> and has already been printed to the console.</p></div>

In [ ]:
corpus = ['The lion is the king of the jungle',
             'Lions have lifespans of a decade', 
             'The lion is an endangered species']

Instructions
<ul>
<li>Instantiate a <code>CountVectorizer</code> object. Name it <code>vectorizer</code>.</li>
<li>Using <code>fit_transform()</code>, generate <code>bow_matrix</code> for <code>corpus</code>.</li>
<li>Using the <code>get_feature_names()</code> method, map the column names to the corresponding word in the vocabulary.</li>
</ul>

In [ ]:
# Create CountVectorizer object
vectorizer = CountVectorizer()

# Generate matrix of word vectors
bow_matrix = vectorizer.fit_transform(corpus)

# Convert bow_matrix into a DataFrame
bow_df = pd.DataFrame(bow_matrix.toarray())

# Map the column names to vocabulary 
bow_df.columns = vectorizer.get_feature_names()

# Print bow_df
bow_df

an  decade  endangered  have  is  ...  lion  lions  of  species  the
0   0       0           0     0   1  ...     1      0   1        0    3
1   0       1           0     1   0  ...     0      1   1        0    0
2   1       0           1     0   1  ...     1      0   0        1    1

[3 rows x 13 columns]

**Observe that the column names refer to the token whose frequency is being recorded. Therefore, since the first column name is an, the first feature represents the number of times the word 'an' occurs in a particular sentence. get_feature_names() essentially gives us a list which represents the mapping of the feature indices to the feature name in the vocabulary.**

## Building a BoW Naive Bayes classifier

### BoW vectors for movie reviews

<div class=""><p>In this exercise, you have been given two pandas Series, <code>X_train</code> and <code>X_test</code>, which consist of movie reviews. They represent the training and the test review data respectively. Your task is to preprocess the reviews and generate BoW vectors for these two sets using <code>CountVectorizer</code>.</p>
<p>Once we have generated the BoW vector matrices <code>X_train_bow</code> and <code>X_test_bow</code>, we will be in a very good position to apply a machine learning model to it and conduct sentiment analysis.</p></div>

In [ ]:
from sklearn.model_selection import train_test_split
df = pd.read_csv('https://github.com/lnunesAI/Datacamp/raw/main/2-machine-learning-scientist-with-python/13-Feature%20Engineering%20for%20NLP%20in%20Python/datasets/movies_reviews_sentiment.csv')
X = df['review']
y = df['sentiment']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42, stratify=y) #0.25

Instructions
<ul>
<li>Import <code>CountVectorizer</code> from the <code>sklearn</code> library.</li>
<li>Instantiate a <code>CountVectorizer</code> object named <code>vectorizer</code>. Ensure that all words are converted to lowercase and <code>english</code> stopwords are removed.</li>
<li>Using <code>X_train</code>, fit <code>vectorizer</code> and then use it to transform <code>X_train</code> to generate the set of BoW vectors <code>X_train_bow</code>.</li>
<li>Transform <code>X_test</code> using <code>vectorizer</code> to generate the set of BoW vectors <code>X_test_bow</code>.</li>
</ul>

In [ ]:
# Import CountVectorizer
from sklearn.feature_extraction.text import CountVectorizer

# Create a CountVectorizer object
vectorizer = CountVectorizer(lowercase=True, stop_words='english')

# Fit and transform X_train
X_train_bow = vectorizer.fit_transform(X_train)

# Transform X_test
X_test_bow = vectorizer.transform(X_test)

# Print shape of X_train_bow and X_test_bow
print(X_train_bow.shape)
print(X_test_bow.shape)

(750, 15027)
(250, 15027)


**You now have a good idea of preprocessing text and transforming them into their bag-of-words representation using CountVectorizer. In this exercise, you have set the lowercase argument to True. However, note that this is the default value of lowercase and passing it explicitly is not necessary. Also, note that both X_train_bow and X_test_bow have 8158 features. There were words present in X_test that were not in X_train. CountVectorizer chose to ignore them in order to ensure that the dimensions of both sets remain the same.**

### Predicting the sentiment of a movie review

<div class=""><p>In the previous exercise, you generated the bag-of-words representations for the training and test movie review data. In this exercise, we will use this model to train a Naive Bayes classifier that can detect the sentiment of a movie review and compute its accuracy. Note that since this is a binary classification problem, the model is only capable of classifying a review as either positive (1) or negative (0). It is incapable of detecting neutral reviews.</p>
<p>In case you don't recall, the training and test BoW vectors are available as <code>X_train_bow</code> and <code>X_test_bow</code> respectively. The corresponding labels are available as <code>y_train</code> and <code>y_test</code> respectively. Also, for you reference, the original movie review dataset is available as <code>df</code>.</p></div>

In [ ]:
from sklearn.naive_bayes import MultinomialNB

Instructions
<ul>
<li>Instantiate an object of <code>MultinomialNB</code>. Name it <code>clf</code>.</li>
<li>Fit <code>clf</code> using <code>X_train_bow</code> and <code>y_train</code>.</li>
<li>Measure the accuracy of <code>clf</code> using <code>X_test_bow</code> and <code>y_test</code>.</li>
</ul>

In [ ]:
# Create a MultinomialNB object
clf = MultinomialNB()

# Fit the classifier
clf.fit(X_train_bow, y_train)

# Measure the accuracy
accuracy = clf.score(X_test_bow, y_test)
print("The accuracy of the classifier on the test set is %.3f" % accuracy)

# Predict the sentiment of a negative review
review = "The movie was terrible. The music was underwhelming and the acting mediocre."
prediction = clf.predict(vectorizer.transform([review]))[0]
print("The sentiment predicted by the classifier is %i" % (prediction))

The accuracy of the classifier on the test set is 0.792
The sentiment predicted by the classifier is 0


In [ ]:
X_train_bow.shape

(750, 14954)

**You have successfully performed basic sentiment analysis. Note that the accuracy of the classifier is 73.2%. Considering the fact that it was trained on only 750 reviews, this is reasonably good performance. The classifier also correctly predicts the sentiment of a mini negative review which we passed into it.**

## Building n-gram models

### n-gram models for movie tag lines

<div class=""><p>In this exercise, we have been provided with a <code>corpus</code> of more than 9000 movie tag lines. Our job is to generate n-gram models up to n equal to 1, n equal to 2 and n equal to 3 for this data and discover the number of features for each model.</p>
<p>We will then compare the number of features generated for each model.</p></div>

In [ ]:
df = pd.read_csv('https://raw.githubusercontent.com/lnunesAI/Datacamp/main/2-machine-learning-scientist-with-python/13-Feature%20Engineering%20for%20NLP%20in%20Python/datasets/movie_overviews.csv').dropna()
df['tagline'] = df['tagline'].str.lower()
corpus = df['tagline']

Instructions
<ul>
<li>Generate an n-gram model with n-grams up to n=1. Name it <code>ng1</code></li>
<li>Generate an n-gram model with n-grams up to n=2. Name it <code>ng2</code></li>
<li>Generate an n-Gram Model with n-grams up to n=3. Name it <code>ng3</code></li>
<li>Print the number of features for each model.</li>
</ul>

In [ ]:
# Generate n-grams upto n=1
vectorizer_ng1 = CountVectorizer(ngram_range=(1,1))
ng1 = vectorizer_ng1.fit_transform(corpus)

# Generate n-grams upto n=2
vectorizer_ng2 = CountVectorizer(ngram_range=(1,2))
ng2 = vectorizer_ng2.fit_transform(corpus)

# Generate n-grams upto n=3
vectorizer_ng3 = CountVectorizer(ngram_range=(1, 3))
ng3 = vectorizer_ng3.fit_transform(corpus)

# Print the number of features for each model
print("ng1, ng2 and ng3 have %i, %i and %i features respectively" % (ng1.shape[1], ng2.shape[1], ng3.shape[1]))

ng1, ng2 and ng3 have 6614, 37100 and 76881 features respectively


**You now know how to generate n-gram models containing higher order n-grams. Notice that ng2 has over 37,000 features whereas ng3 has over 76,000 features. This is much greater than the 6,000 dimensions obtained for ng1. As the n-gram range increases, so does the number of features, leading to increased computational costs and a problem known as the curse of dimensionality.**

### Higher order n-grams for sentiment analysis

<div class=""><p>Similar to a previous exercise, we are going to build a classifier that can detect if the review of a particular movie is positive or negative. However, this time, we will use n-grams up to n=2 for the task.</p>
<p>The n-gram training reviews are available as <code>X_train_ng</code>. The corresponding test reviews are available as <code>X_test_ng</code>. Finally, use <code>y_train</code> and <code>y_test</code> to access the training and test sentiment classes respectively.</p></div>

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.5, random_state=42, stratify=y)
ng_vectorizer = CountVectorizer(ngram_range=(1, 2))
X_train_ng = ng_vectorizer.fit_transform(X_train)
X_test_ng = ng_vectorizer.transform(X_test)

Instructions
<ul>
<li>Define an instance of MultinomialNB. Name it <code>clf_ng</code></li>
<li>Fit the classifier on <code>X_train_ng</code> and <code>y_train</code>.</li>
<li>Measure <code>accuracy</code> on <code>X_test_ng</code> and <code>y_test</code> the using <code>score()</code> method.</li>
</ul>

In [ ]:
# Define an instance of MultinomialNB 
clf_ng = MultinomialNB()

# Fit the classifier 
clf_ng.fit(X_train_ng, y_train)

# Measure the accuracy 
accuracy = clf_ng.score(X_test_ng, y_test)
print("The accuracy of the classifier on the test set is %.3f" % accuracy)

# Predict the sentiment of a negative review
review = "The movie was not good. The plot had several holes and the acting lacked panache."
prediction = clf_ng.predict(ng_vectorizer.transform([review]))[0]
print("The sentiment predicted by the classifier is %i" % (prediction))

The accuracy of the classifier on the test set is 0.758
The sentiment predicted by the classifier is 0


**You're now adept at performing sentiment analysis using text. Notice how this classifier performs slightly better than the BoW version. Also, it succeeds at correctly identifying the sentiment of the mini-review as negative. In the next chapter, we will learn more complex methods of vectorizing textual data.**

### Comparing performance of n-gram models

<div class=""><p>You now know how to conduct sentiment analysis by converting text into various n-gram representations and feeding them to a classifier. In this exercise, we will conduct sentiment analysis for the same movie reviews from before using two n-gram models: unigrams and n-grams upto n equal to 3. </p>
<p>We will then compare the performance using three criteria: accuracy of the model on the test set, time taken to execute the program and the number of features created when generating the n-gram representation.</p></div>

In [ ]:
import time

Instructions 1/2
<p>Initialize a CountVectorizer object such that it generates unigrams.</p>

In [ ]:
start_time = time.time()
# Splitting the data into training and test sets
train_X, test_X, train_y, test_y = train_test_split(X, y, test_size=0.5, random_state=42, stratify=y)

# Generating ngrams
vectorizer = CountVectorizer()
train_X = vectorizer.fit_transform(train_X)
test_X = vectorizer.transform(test_X)

# Fit classifier
clf = MultinomialNB()
clf.fit(train_X, train_y)

# Print accuracy, time and number of dimensions
print("The program took %.3f seconds to complete. The accuracy on the test set is %.2f. The ngram representation had %i features." % (time.time() - start_time, clf.score(test_X, test_y), train_X.shape[1]))

The program took 0.209 seconds to complete. The accuracy on the test set is 0.75. The ngram representation had 12347 features.


Instructions 2/2
<p>Initialize a CountVectorizer object such that it generates ngrams upto n=3.</p>

In [ ]:
# Splitting the data into training and test sets
train_X, test_X, train_y, test_y = train_test_split(X, y, test_size=0.5, random_state=42, stratify=y)

# Generating ngrams
vectorizer = CountVectorizer(ngram_range=(1, 3))
train_X = vectorizer.fit_transform(train_X)
test_X = vectorizer.transform(test_X)

# Fit classifier
clf = MultinomialNB()
clf.fit(train_X, train_y)

# Print accuracy, time and number of dimensions
print("The program took %.3f seconds to complete. The accuracy on the test set is %.2f. The ngram representation had %i features." % (time.time() - start_time, clf.score(test_X, test_y), train_X.shape[1]))

The program took 40.127 seconds to complete. The accuracy on the test set is 0.77. The ngram representation had 178240 features.


**The program took around 0.2 seconds in the case of the unigram model and more than 10 times longer for the higher order n-gram model. The unigram model had over 12,000 features whereas the n-gram model for upto n=3 had over 178,000! Despite taking higher computation time and generating more features, the classifier only performs marginally better in the latter case, producing an accuracy of 77% in comparison to the 75% for the unigram model.**